src/lib/util/pair_tests.c

```diff
@@ -0,0 +1,1205 @@
+
+		attr = fr_dict_attr_by_name(NULL, parent, ctx->name);
+		if (!attr) return -1;
+
+		/* Set the VALUES */
+		if (ctx->values) {
+			fr_dict_adhoc_attr_value_t *v;
+
+			for (v = ctx->values;
+			     v->key != NULL;
+			     v++) fr_dict_attr_enum_add_name(fr_dict_attr_unconst(attr), v->key, &v->val, false, false);
+		}
+
+		*ctx->da = attr;
+	}
+
+	return 0;
+}
+
+static void pair_tests_init(void)
+{
+	autofree = talloc_autofree_context();
+	if (!autofree) {
+	error:
+		fr_perror("pair_tests");
+		fr_exit_now(EXIT_FAILURE);
+	}
+
+	/*
+	 *	Mismatch between the binary and the libraries it depends on
+	 */
+	if (fr_check_lib_magic(RADIUSD_MAGIC_NUMBER) < 0) goto error;
+
+	if (!fr_dict_global_ctx_init(autofree, dict_dir)) goto error;
+
+	/*
+	 *	Set the root name of the dictionary
+	 */
+	dict_test = fr_dict_alloc("test", 666);
+	if (!dict_test) goto error;
+
+	if (init_adhoc_attrs(test_dict_attrs) < 0) goto error;
+
+	/* Initialize the "sample_pairs" list */
+	if (load_attr_pairs(&sample_pairs) < 0) goto error;
+}
+
+/*
+ *	Tests functions
+ */
+static void test_fr_pair_afrom_da(void)
+{
+	fr_pair_t   *vp;
+
+	TEST_CASE("Allocation using fr_pair_afrom_da");
+	TEST_CHECK((vp = fr_pair_afrom_da(autofree, attr_test_integer)) != NULL);
+
+	TEST_CASE("Validating VP_VERIFY()");
+	VP_VERIFY(vp);
+
+	vp->vp_uint32 = 12345;
+
+	TEST_CHECK(vp->vp_uint32 == 12345);
+	TEST_MSG("Expected %s == 12345", attr_test_integer->name);
+
+	talloc_free(vp);
+}
+
+static void test_fr_pair_afrom_child_num(void)
+{
+	fr_pair_t    *vp;
+	unsigned int attr = FR_MODULE_RETURN_CODE;
+
+	TEST_CASE("Allocation using fr_pair_afrom_child_num");
+	TEST_CHECK((vp = fr_pair_afrom_child_num(autofree, fr_dict_root(dict_test), attr)) != NULL);
+
+	TEST_CASE("Validating VP_VERIFY()");
+	VP_VERIFY(vp);
+
+	TEST_CHECK(vp && vp->da && vp->da->attr == FR_MODULE_RETURN_CODE);
+	TEST_MSG("Expected attr(%d) == vp->da->attr(%d)", attr, vp->da->attr);
+
+	talloc_free(vp);
+}
+
+static void test_fr_pair_copy(void)
+{
+	fr_pair_t   *vp, *copy;
+
+	TEST_CASE("Allocation using fr_pair_copy");
+	TEST_CHECK((vp = fr_pair_afrom_da(autofree, attr_test_integer)) != NULL);
+
+	TEST_CASE("Validating VP_VERIFY()");
+	VP_VERIFY(vp);
+
+	copy = fr_pair_copy(autofree, vp);
+
+	TEST_CASE("Validating VP_VERIFY()");
+	VP_VERIFY(copy);
+
+	TEST_CHECK(copy && copy->da && copy->da->type == FR_TYPE_UINT32);
+	TEST_MSG("Expected (copy->type == FR_TYPE_UINT32)");
+
+	talloc_free(vp);
+	talloc_free(copy);
+}
+
+static void test_fr_pair_steal(void)
+{
+	fr_pair_t   *vp;
+	TALLOC_CTX 	*ctx = talloc_null_ctx();
+
+	TEST_CASE("Allocate a new attribute fr_pair_afrom_da");
+	TEST_CHECK((vp = fr_pair_afrom_da(ctx, attr_test_integer)) != NULL);
+
+	TEST_CASE("Validating VP_VERIFY()");
+	VP_VERIFY(vp);
+
+	TEST_CASE("Stealing 'vp' pair using fr_pair_steal()");
+	fr_pair_steal(autofree, vp); /* It should exit without memory-leaks */
```

Comment:
Yes, but you need to test the steal was performed using talloc_parent() to check the parent...

Reply:
Good catch. Fixed.